server/src/main/java/org/apache/iotdb/db/query/reader/resourceRelated/UnseqResourceMergeReader.java

```diff
@@ -42,136 +42,130 @@
-      if (!metaDataList.isEmpty()) {
+      if (!tsFileMetaDataList.isEmpty()) {
         TsFileSequenceReader tsFileReader = FileReaderManager.getInstance()
             .get(tsFileResource, tsFileResource.isClosed());
-        chunkLoader = new ChunkLoaderImpl(tsFileReader);
-      }
+        ChunkLoaderImpl chunkLoader = new ChunkLoaderImpl(tsFileReader);
 
-      for (ChunkMetaData chunkMetaData : metaDataList) {
-
-        if (timeFilter != null && !timeFilter.satisfy(chunkMetaData.getStatistics())) {
-          continue;
+        for (ChunkMetaData chunkMetaData : tsFileMetaDataList) {
+          if (timeFilter != null && !timeFilter.satisfy(chunkMetaData.getStatistics())) {
+            tsFileMetaDataList.remove(chunkMetaData);
+            continue;
+          }
+          chunkMetaData.setPriority(priority++);
+          chunkMetaData.setChunkLoader(chunkLoader);
         }
-        // create and add DiskChunkReader
-        readerWrapList.add(new ChunkReaderWrap(chunkMetaData, chunkLoader, timeFilter));
+        metaDataList.addAll(tsFileMetaDataList);
       }
 
+      // create and add MemChunkReader
       if (!tsFileResource.isClosed()) {
-        // create and add MemChunkReader
-        readerWrapList.add(new ChunkReaderWrap(tsFileResource.getReadOnlyMemChunk(), timeFilter));
+        ChunkReaderWrap memChunkReaderWrap = new ChunkReaderWrap(
+            tsFileResource.getReadOnlyMemChunk(), timeFilter);
+        priorityMergeReader.addReaderWithPriority(memChunkReaderWrap.getIPointReader(), priority++);
       }
     }
 
-    ExternalSortJobEngine externalSortJobEngine = SimpleExternalSortEngine.getInstance();
-    List<IPointReader> readerList = externalSortJobEngine
-        .executeForIPointReader(queryId, readerWrapList);
-    int priorityValue = 1;
-    for (IPointReader chunkReader : readerList) {
-      addReaderWithPriority(chunkReader, priorityValue++);
-    }
+    // sort All ChunkMetadata by start time
+    metaDataList = metaDataList.stream().sorted(Comparator.comparing(ChunkMetaData::getStartTime))
+        .collect(Collectors.toList());
   }
 
-
   /**
-   * Zesong Sun !!!
+   * Create a ChunkReader with priority for each ChunkMetadata and put the ChunkReader to
+   * mergeReader one by one
    */
   @Override
   public boolean hasNextBatch() throws IOException {
-    return false;
-  }
+    ChunkMetaData metaData;
+    ChunkReaderWrap diskChunkReaderWrap;
+    if (priorityMergeReader.hasNext()) {
+      metaData = metaDataList.get(index);
+      // create and add DiskChunkReader
+      while (index < metaDataList.size()) {
+        diskChunkReaderWrap = new ChunkReaderWrap(metaData, metaData.getChunkLoader(), timeFilter);
+        priorityMergeReader
+            .addReaderWithPriority(diskChunkReaderWrap.getIPointReader(), metaData.getPriority());
+        metaData = metaDataList.get(index++);
+        if (priorityMergeReader.current().getTime() < metaData.getStartTime()) {
+          break;
+        }
```

Comment:
Thanks for your quick reaction. 
But ... the modified logic senses more strange to me, especially the two places:
1) The `index++` logic makes the last `metaData` of `metaDataList` got at line 133 can never be added when its `getStartTime() <= priorityMergeReader.current().getTime()`, because the program will exit in the next loop at line 129.
2) the `addReaderWithPriority` is directly used before applying the `if (priorityMergeReader.current().getTime() < metaData.getStartTime())` check.

You can modify or discuss about this later. Take your time :)

Reply:
Sorry for my fault... I have changed the order, how do you think about the codes now? 😂